update template for adding a database connection

arunge opened the pull request:
#1 
- ~update the template to add a database~
- ~add a template to describe the format for new environment variables~

as discussed in https://github.com/Pandora-IsoMemo/iso-data/pull/2#discussion_r850417485

- two new functions `createNewDBSource()` and `createNewDBSource()` to automatically 
  - create a new script `R/02-<datasource>.R`
  - add an entry to `R/00-databases.R`
  - update the `.Renviron` for mysql database sources
- update of the README.md with the new description how to add a new source
- tests for the new functions

wahani commented on R/02-template-Renviron.R:
```diff
@@ -0,0 +1,13 @@
+# This template contains the format of the environment variables that are mentioned in the template:
```
The .Renviron should be at the root of the R session and would be ignored otherwise. Maybe we should then place the template at root also?

I think a better option would be to just document this in the template-db.R and in the documentation. Hence don't have this as template file.

Another Option would be to generate these files, instead of providing templates. This would be in the spirit of [usethis](https://cran.r-project.org/web/packages/usethis/index.html) and maybe much easier because we can add entries to the ..Renviron, make sure it is in the gitignore, and properly setup the database template.

arunge replied:
As just discussed, I will go for the second option and prepare a function that adjusts all which is required to add a database.
@wahani Thanks for this valuable feedback.

arunge commented:
@wahani Can you review the changes for template creation: https://github.com/Pandora-IsoMemo/iso-data/pull/2#issue-1191895591 ?

arunge commented:
@wahani There is only one thing left. The README names two (former) branches under the section "Deployment" (https://github.com/Pandora-IsoMemo/iso-data/blob/main/README.md#deployment). Can you update this small part such that it reflects the current setup?

wahani commented on R/01-createNewDBSource.R:
```diff
@@ -0,0 +1,168 @@
+#' @param tableName name of the table containing the data
+#' @param descriptionCreator (character) command that creates the description, e.g. pasting data
+#'  columns "var1" and "var2": "paste(isoData$var1, isoData$var2)"
+#' @param datingType (character) dating type for the database
+#' @param coordType (character) coord type for the database
```
I don't understand those two arguments. Maybe this is obvious with more domain knowledge. Maybe an example would help. Like what are the datingTypes / coordTypes for existing data sources.

arunge replied:
I updated the description with some examples. But I found nothing in this package or the app package regarding these parameters.

wahani commented on R/01-createNewDBSource.R:
```diff
@@ -0,0 +1,168 @@
+                              coordType,
+                              descriptionCreator = NULL,
+                              scriptFolder = "R",
+                              rootFolder = "") {
```
```suggestion
                              rootFolder = ".") {
```

wahani commented on R/01-createNewDBSource.R:
```diff
@@ -0,0 +1,168 @@
+#' @param coordType (character) coord type for the database
+#' @param scriptFolder (character) place to store the scripts.
+#' @param rootFolder (character) root folder of the package, usually containing .Renviron,
+#' DESCRIPTION, ...
```
The function is not exported and visible to a user.

wahani commented on R/01-createNewDBSource.R:
```diff
@@ -0,0 +1,168 @@
+#' Paste DB settings
+#'
+#' @inheritParams createNewDBSource
+addDBSettings <- function(dbName, tableName) {
```
I find these sort of tasks much easier to read and understand when we use a templating system. For dbtools and SQL Queries we use templates. E.g. see https://cran.r-project.org/web/packages/templates/vignettes/Template_Programming_in_R.html

I am the author of dbtools and templates and obviously biased, please make a change only if you see it as an improvement.

In case you want to use templates, you may also want to put the template into its own file. Like `./inst/templates/db-settings.R`. You can then reliably refer to that file with `system.file`.

arunge replied:
Now, I am using templates. Thanks for that remark @wahani ! I added 3 new templates to `./inst/templates`

wahani reviewed (changes requested):
Currently I get an error. Maybe reproducible with

```
> library(MpiIsoData)
> MpiIsoData:::createNewDBSource(
+     "my-database",
    "my-table",
+     "my-table",
+     "dating-type",
+     "coord-type"
+ )
Fehler in file(con, "w") : kann Verbindung nicht öffnen
Zusätzlich: Warnmeldung:
In file(con, "w") :
  kann Datei '/.Renviron' nicht öffnen: Keine Berechtigung
```

Note that I have already an .Renviron in my working directory. I also **do not** load the package via devtools::load_all().

I guess it would also make sense to validate the input like `my-database` because it needs to be valid R code. E.g. with an regex of the form `grepl("^[[:alnum:]]+$", argument)`


Mybe we should add some user output which files are

- created
- added

To make that behavior transparent even if git is not being used.

wahani commented on R/01-createNewDBSource.R:
```diff
@@ -0,0 +1,168 @@
+#'
+#' @param dbName (character) name of the new database source, e.g. "xyDBname"
+#' @param tableName name of the table containing the data
+#' @param descriptionCreator (character) command that creates the description, e.g. pasting data
+#'  columns "var1" and "var2": "paste(isoData$var1, isoData$var2)"
```
I think it would be more transparent if you define it in the extract method and remove the parameter.

arunge replied:
I removed the parameter and made some comments in the automatically generated draft.

arunge commented:
> Currently I get an error. Maybe reproducible with
> 
> ```
> > library(MpiIsoData)
> > MpiIsoData:::createNewDBSource(
> +     "my-database",
>     "my-table",
> +     "my-table",
> +     "dating-type",
> +     "coord-type"
> + )
> Fehler in file(con, "w") : kann Verbindung nicht öffnen
> Zusätzlich: Warnmeldung:
> In file(con, "w") :
>   kann Datei '/.Renviron' nicht öffnen: Keine Berechtigung
> ```
> 
> Note that I have already an .Renviron in my working directory. I also **do not** load the package via devtools::load_all().

This was not reproducable, but I already updated: `rootFolder = "."` Maybe, this was the problem.

> I guess it would also make sense to validate the input like `my-database` because it needs to be valid R code. E.g. with an regex of the form `grepl("^[[:alnum:]]+$", argument)`

Now the `dbName` will be formated before usage. Is this ok?
```R
formatDBName <- function(dbName) {
  # upper letters
  res <- toupper(dbName)
  # replace special characters with underscore
  res <- gsub("[^[:alnum:]]", "_", res)
  # replace "__" with "_"
  gsub("_+", "_", res)
}
```

> 
> Mybe we should add some user output which files are
> 
> * created
> * added
> 
> To make that behavior transparent even if git is not being used.

I added logging functions, e.g. `logging("Creating new file: %s", filePath)`

wahani reviewed (approved):
Looks good now. 

- Don't get any errors anymore.
- Seems to be doing what we need.
- Code is easy to read.
- Documentation is accessible and exports are present.
